### PR TITLE
Make cram tests available outside of dune

### PR DESCRIPTION
**Pull request by diml**
This PR transform our internal `cram` tool into a general purpose public `dune-cram` tool distributed as a new `dune-cram` package.

To that purpose, this PR cleans up and generalises the sanitiser as follow.

## Rewriting of ocamlc_flag and other vars

This PR get rid of this rewriting. It was only used to rewrite the output of `dune printenv`. Instead, this PR:
- adds an option `--field <name>` to `dune printenv` to select the field to print
- use this option in tests to limit the output in some cases
- explicitly set the default `c_flags` in the test of the `c_flags` env field

## Rewriting of os-specific file extensions

This is kept as it

## Rewriting of the pwd

This is kept, but generalised as follow: the sanitizer rewrite absolute paths using the standard `BUILD_PATH_PREFIX_MAP` environment variable. It does so using the [build_path_prefix_map library](https://gitlab.com/gasche/build_path_prefix_map) which this PR vendors. This look like something we are going to need at some point anyway to improve reproducibility.

Tests are allowed to extend this variable in order to rewrite more paths, as is done by the `install-libdir` test. This is a nice way to rewrite paths in a fully portable way.

##  Next steps

- Write doc
- Add a `cram` stanza to make writing cram test even easier

### CODE REVIEW

**Comment by diml** [comment]
The only thing a tad unsatisfying about this PR is that the cram tool is pretty much independent of dune, however the rewriting of file extensions is tied to running inside Dune. Indeed, we obtained the various file extension (`ext_exe`, `ext_obj`, ...) from `ocamlc -config`.

Not sure what we could do about this. When we introduce the `cram` stanza, we could make this explicit, for instance by listing the extensions to rewrite:

```
(cram
 ...
 (sanitize_file_extensions
  (%{ext_obj} $ext_obj)
  (%{ext_dll} $ext_dll)))
```
It's a tad more verbose, but at least what's happening is explicit.

**Comment by emillon** [comment]
Hi!

First, thanks for this, it's one part of dune that I miss when working with other projects.

> Indeed, we obtained the various file extension (`ext_exe`, `ext_obj`, ...) from `ocamlc -config`.
> 
> Not sure what we could do about this.

Could we run and parse `ocamlc -config` within cram itself?

**Comment by diml** [comment]
No problem :)

> Could we run and parse ocamlc -config within cram itself?

That's basically what it does now (except that dune passes the output of `ocamlc -config` via a file). However, it feels a bit ad-hoc. `dune-cram` is completely independent of the OCaml toolchain apart from this detail. Additionally, there is also a versioning issue. For instance, in the future we might want to rewrite more extensions and that would break backward compatibility.

All this makes me think that the list of extension to rewrite should be part of the configuration of the tool rather than builtin.

For instance, we could add an argument `--rewrite-extension <ext>=<repl>`. Actually, that seems best and we can just add a small wrapper to not have to type this everytime. I'm going to try that.

**Comment by emillon** [comment]
Sounds good. The set of rewritten extensions can be versioned in dune itself then.
By the way, how specific is this to extensions? does it consider only those that appear it filenames, or is that a generic text replacing?

**Comment by diml** [comment]
It's (almost) generic text replacing. The regexp is: a character other than `/` followed by the extension. Ideally it should be just filenames, though I'm not too sure what regexp to use here...

**Comment by diml** [comment]
As discussed during last week meeting, instead of doing this extension rewriting afterwards, we should just add a "portable output" mode to dune and enable this mode by default when running dune inside dune. That will side-step this issue entirely.

I'd still like to merge this PR now though as it improves the current codebase and touches a large number of files. So I removed the `dune-cram` package from this PR and made the binary private again.

File: otherlibs/cram/bin/dune [hunk 1]
```diff
@@ -0,0 +1,11 @@
+(executable
+ (name dune_cram)
+ (libraries dune-private-libs.stdune
+            dune-build-info
+            cmdliner
+            build_path_prefix_map
+            dune-private-libs.dune_re
```
Comment (rgrinberg) [inline otherlibs/cram/bin/dune]: Perhaps the publicly released version of cram should use cmdliner and re from opam. That would teach more about some of the bundling related problem we're trying to solve :)

Comment (diml) [reply otherlibs/cram/bin/dune]: Haha maybe. But that's a problem for later :)

**Pull request merged**